src/backend/cdb/motion/test/cdbsenddummypacket_test.c

```diff
@@ -148,10 +150,95 @@ create_sender_socket(sa_family_t af)
 	return sockfd;
 }
 
+static bool
+check_for_ipv6(void)
+{
+	const struct sockaddr_in6 test_addr = {
+		.sin6_family = AF_INET6,
+		.sin6_addr = in6addr_loopback
```

Comment:
why are we checking on the loopback interface? The test passes without changes in ic_udpifc.c if ipv6 is enabled only on the loopback interface, is this the expected behavior?

Reply:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/freeaddrinfo.html

> If the AI_PASSIVE flag is specified, the returned address information shall be suitable for use in binding a socket for accepting incoming connections for the specified service. In this case, if the nodename argument is null, then the IP address portion of the socket address structure shall be set to INADDR_ANY for an IPv4 address or IN6ADDR_ANY_INIT for an IPv6 address. If the AI_PASSIVE flag is not specified, the returned address information shall be suitable for a call to [connect()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html) (for a connection-mode protocol) or for a call to [connect()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html), [sendto()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sendto.html), or [sendmsg()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sendmsg.html) (for a connectionless protocol). In this case, if the nodename argument is null, then the IP address portion of the socket address structure shall be set to the loopback address. The AI_PASSIVE flag shall be ignored if the nodename argument is not null.

We pass `AI_PASSIVE`, if node is `NULL`:
https://github.com/arenadata/gpdb/blob/cbdcfc62d9b6ef1462c609d56dafad8d133ccb47/src/backend/cdb/motion/ic_udpifc.c#L1226-L1227

That means we are calling `getaddrinfo()` with `in6addr_any`. `man` says:
> If hints.ai_flags includes the AI_ADDRCONFIG flag, then IPv6 addresses are returned only if the local system has at least one IPv6 address configured.  The loopback address is not considered for this case as valid as a configured address.

So you are right, and this will be a edge case when IPv6 is only enabled on a loopback device. We can check for IPv6 support excluding loopback devices with `getifaddrs()` instead.